File: android/app/src/main/kotlin/ai/gowtam/oak/services/DexLookupService.kt
package ai.gowtam.oak.services

import ai.gowtam.oak.networking.Endpoint
import ai.gowtam.oak.networking.OakApiClient
import ai.gowtam.oak.networking.OakError
import ai.gowtam.oak.wire.DexSpriteRef
import ai.gowtam.oak.wire.EntityKind
import ai.gowtam.oak.wire.Format
import ai.gowtam.oak.wire.LearnsetMove
import ai.gowtam.oak.wire.SearchMatch
import android.util.Log
import kotlinx.serialization.Serializable

/**
 * The team-builder entity-picker seam: typeahead search, per-species learnsets, and
 * batch sprite/type/ability/base-stat refs (component-design.md "Services layer";
 * mirrors iOS `DexLookupService`). Backs `TeamEditorViewModel`'s species / ability /
 * item / move pickers so it unit-tests against a fake.
 *
 * All three routes are public, read-only Pokédex reference data (no Bearer header)
 * and **never throw** — a picker that can't reach the index just shows no
 * suggestions / no sprite rather than breaking the editor ([ArtifactService]'s same
 * never-throw policy).
 */
interface DexLookupService {
    /**
     * `GET /api/search` — ranked typeahead candidates for one entity [kind] in
     * [format]. A blank [query] returns an alphabetical browse listing (mirrors the
     * route's blank-query behavior); any fault folds to `[]`.
     */
    suspend fun search(kind: EntityKind, query: String, format: Format): List<SearchMatch>

    /**
     * `GET /api/learnset` — the legal movepool for [pokemon] in [format], sorted by
     * display name. An unknown species or any fault folds to `[]`.
     */
    suspend fun learnset(pokemon: String, format: Format): List<LearnsetMove>

    /**
     * `GET /api/sprites` — batch sprite/type/ability/base-stat refs for [names]
     * (species slugs) in [format], keyed by the requested name. Unknown names are
     * simply absent; any fault folds to `{}`. An empty [names] short-circuits to `{}`
     * with no request.
     */
    suspend fun sprites(names: List<String>, format: Format): Map<String, DexSpriteRef>
}

/**
 * Production [DexLookupService] over [OakApiClient]. Every method catches the
 * [OakError] the client throws and maps it to the documented empty result, logging
 * only a non-sensitive label (never the query text).
 */
class LiveDexLookupService(private val apiClient: OakApiClient) : DexLookupService {

    override suspend fun search(kind: EntityKind, query: String, format: Format): List<SearchMatch> {
        val endpoint = Endpoint(
            method = Endpoint.Method.GET,
            path = "/api/search",
            queryItems = listOf(
                "kind" to kind.wireValue,
                "q" to query,
                "format" to format.rawValue,
            ),
            requiresAuth = false,
        )
        return try {
            apiClient.send(endpoint, SearchEnvelope.serializer()).matches
        } catch (e: OakError) {
            Log.e(TAG, "entity search unavailable (kind ${kind.wireValue}): ${e::class.simpleName}")
            emptyList()
        }
    }

    override suspend fun learnset(pokemon: String, format: Format): List<LearnsetMove> {
        val endpoint = Endpoint(
            method = Endpoint.Method.GET,
            path = "/api/learnset",
            queryItems = listOf(
                "pokemon" to pokemon,
                "format" to format.rawValue,
            ),
            requiresAuth = false,
        )
        return try {
            apiClient.send(endpoint, LearnsetEnvelope.serializer()).moves
        } catch (e: OakError) {
            Log.e(TAG, "learnset unavailable: ${e::class.simpleName}")
            emptyList()
        }
    }

    override suspend fun sprites(names: List<String>, format: Format): Map<String, DexSpriteRef> {
        if (names.isEmpty()) return emptyMap()
        val endpoint = Endpoint(
            method = Endpoint.Method.GET,
            path = "/api/sprites",
            queryItems = listOf(
                "format" to format.rawValue,
                "names" to names.joinToString(","),
            ),
            requiresAuth = false,
        )
        return try {
            apiClient.send(endpoint, SpritesEnvelope.serializer()).refs
        } catch (e: OakError) {
            Log.e(TAG, "sprite batch unavailable: ${e::class.simpleName}")
            emptyMap()
        }
    }

    private companion object {
        const val TAG = "Oak.DexLookup"
    }
}

/** `GET /api/search` → `{ matches: SearchMatch[] }`. */
@Serializable
private data class SearchEnvelope(val matches: List<SearchMatch>)

/** `GET /api/learnset` → `{ moves: LearnsetMove[] }`. */
@Serializable
private data class LearnsetEnvelope(val moves: List<LearnsetMove>)

/** `GET /api/sprites` → `{ refs: { [name]: DexSpriteRef } }`. */
@Serializable
private data class SpritesEnvelope(val refs: Map<String, DexSpriteRef>)

/**
 * A never-fetching [DexLookupService]: every call resolves to an empty result with no
 * network access. A default for call sites that only exercise team CRUD and don't
 * care about pickers; production wiring ([ai.gowtam.oak.app.ServiceContainer]) always
 * passes [LiveDexLookupService] explicitly.
 */
class EmptyDexLookupService : DexLookupService {
    override suspend fun search(kind: EntityKind, query: String, format: Format): List<SearchMatch> = emptyList()
    override suspend fun learnset(pokemon: String, format: Format): List<LearnsetMove> = emptyList()
    override suspend fun sprites(names: List<String>, format: Format): Map<String, DexSpriteRef> = emptyMap()
}

File: ios/OakApp/Services/DexLookupService.swift
import Foundation

/// The team-builder entity-picker seam: typeahead search, per-species learnsets, and
/// batch sprite/type/ability/base-stat refs (component-design.md "Services layer"). Backs
/// ``TeamEditorViewModel``'s species / ability / item / move pickers so it unit-tests
/// against a fake, never the `Live…` concrete.
///
/// All three routes are public, read-only Pokédex reference data (no Bearer header) and
/// **never throw** — mirroring the web client helpers (`search-client.ts` /
/// `learnset-client.ts` / `sprites-client.ts`, all of which fold a transport/HTTP/decode
/// fault to an empty result rather than surfacing an error). A picker that can't reach the
/// index just shows no suggestions / no sprite rather than breaking the editor
/// (`ArtifactService`'s same never-throw policy, ADR-8).
protocol DexLookupService: Sendable {
  /// `GET /api/search` — ranked typeahead candidates for one entity `kind` in `format`. A
  /// blank `query` returns an alphabetical browse listing (mirrors the route's blank-query
  /// behavior); any fault folds to `[]`.
  func search(kind: EntityKind, query: String, format: Format) async -> [SearchMatch]

  /// `GET /api/learnset` — the legal movepool for `pokemon` in `format`, sorted by display
  /// name. An unknown species or any fault folds to `[]`.
  func learnset(pokemon: String, format: Format) async -> [LearnsetMove]

  /// `GET /api/sprites` — batch sprite/type/ability/base-stat refs for `names` (species
  /// slugs) in `format`, keyed by the requested name. Unknown names are simply absent;
  /// any fault folds to `[:]`. An empty `names` short-circuits to `[:]` with no request.
  func sprites(names: [String], format: Format) async -> [String: DexSpriteRef]
}

// MARK: - Live implementation

/// Production ``DexLookupService`` over ``OakAPIClient``. A value type holding one
/// immutable actor reference, so it is `Sendable` without ceremony. Every method catches
/// the ``OakError`` the client throws and maps it to the documented empty result, logging
/// only a non-sensitive label (never the query text — conventions.md "Logging").
struct LiveDexLookupService: DexLookupService {
  private let apiClient: OakAPIClient

  init(apiClient: OakAPIClient) {
    self.apiClient = apiClient
  }

  func search(kind: EntityKind, query: String, format: Format) async -> [SearchMatch] {
    let endpoint = Endpoint(
      method: .get,
      path: "/api/search",
      queryItems: [
        URLQueryItem(name: "kind", value: kind.rawValue),
        URLQueryItem(name: "q", value: query),
        URLQueryItem(name: "format", value: format.rawValue),
      ],
      requiresAuth: false
    )
    do {
      return try await apiClient.send(endpoint, as: SearchEnvelope.self).matches
    } catch {
      Log.network.error("entity search unavailable (kind \(kind.rawValue, privacy: .public))")
      return []
    }
  }

  func learnset(pokemon: String, format: Format) async -> [LearnsetMove] {
    let endpoint = Endpoint(
      method: .get,
      path: "/api/learnset",
      queryItems: [
        URLQueryItem(name: "pokemon", value: pokemon),
        URLQueryItem(name: "format", value: format.rawValue),
      ],
      requiresAuth: false
    )
    do {
      return try await apiClient.send(endpoint, as: LearnsetEnvelope.self).moves
    } catch {
      Log.network.error("learnset unavailable")
      return []
    }
  }

  func sprites(names: [String], format: Format) async -> [String: DexSpriteRef] {
    guard !names.isEmpty else { return [:] }
    let endpoint = Endpoint(
      method: .get,
      path: "/api/sprites",
      queryItems: [
        URLQueryItem(name: "format", value: format.rawValue),
        URLQueryItem(name: "names", value: names.joined(separator: ",")),
      ],
      requiresAuth: false
    )
    do {
      return try await apiClient.send(endpoint, as: SpritesEnvelope.self).refs
    } catch {
      Log.network.error("sprite batch unavailable")
      return [:]
    }
  }
}

// MARK: - Wire envelopes (private to the service)

/// `GET /api/search` → `{ matches: SearchMatch[] }`.
private struct SearchEnvelope: Decodable, Sendable {
  let matches: [SearchMatch]
}

/// `GET /api/learnset` → `{ moves: LearnsetMove[] }`.
private struct LearnsetEnvelope: Decodable, Sendable {
  let moves: [LearnsetMove]
}

/// `GET /api/sprites` → `{ refs: { [name]: DexSpriteRef } }`.
private struct SpritesEnvelope: Decodable, Sendable {
  let refs: [String: DexSpriteRef]
}

// MARK: - Null object (default for call sites that don't need dex data)

/// A never-fetching ``DexLookupService``: every call resolves to an empty result with no
/// network access. This is ``TeamEditorViewModel``/``TeamsListViewModel``'s default
/// argument, so the many existing call sites that only exercise team CRUD (and don't care
/// about pickers) keep compiling unchanged; production wiring (``ServiceContainer``)
/// always passes ``LiveDexLookupService`` explicitly.
struct EmptyDexLookupService: DexLookupService {
  func search(kind: EntityKind, query: String, format: Format) async -> [SearchMatch] { [] }
  func learnset(pokemon: String, format: Format) async -> [LearnsetMove] { [] }
  func sprites(names: [String], format: Format) async -> [String: DexSpriteRef] { [:] }
}
